hw/dv/sv/dv_base_reg/csr_excl_item.sv

```
@@ -9,103 +9,145 @@ class csr_excl_item extends uvm_object;
+  // arg csr_excl_type: The desired exclusion type.
+  // arg csr_test_type: The desired test type for which the exclusion is in effect.
   virtual function void add_excl(string obj,
                                  csr_excl_type_e csr_excl_type,
                                  csr_test_type_e csr_test_type = CsrAllTests);
     bit [2:0] val = CsrNoExcl;
     bit [NUM_CSR_TESTS-1:0] test = CsrInvalidTest;
 
-    if (csr_test_type == CsrInvalidTest) begin
-      `uvm_fatal(`gfn, $sformatf("add %s exclusion without a test", obj))
+    `DV_CHECK_NE_FATAL(csr_test_type, CsrInvalidTest,
+                       $sformatf("Test type not specified for the exclusion of %0s.", obj))
+
+    if (!exclusions.exists(obj)) begin
+      exclusions[obj] = '{enable:1'b1, csr_test_type:csr_test_type, csr_excl_type:csr_excl_type};
+      return;
     end
 
-    if (!exclusions.exists(obj)) exclusions[obj] = '{default:CsrNoExcl};
     val = csr_excl_type | exclusions[obj].csr_excl_type;
     test = csr_test_type | exclusions[obj].csr_test_type;
     exclusions[obj].csr_excl_type = csr_excl_type_e'(val);
     exclusions[obj].csr_test_type = test;
   endfunction
 
-  // function to check if given blk / csr or field AND its parent has been excluded with the
-  // supplied exclusion type
-  // arg uvm_object obj: given blk, csr or field
-  // arg csr_excl_type_e csr_excl_type: exclusion type
+  // Turns exclusion on or off for a block, register or field.
+  //
+  // The originally set exclusions are untouched. This method only enables or disables the
+  // application of the exclusions temporarily. Vacuously returns if no exclusions were found
+  // to begin with.
+  // obj: The hierarchical path to block, register or field.
+  // enable: Bit indicating whether to enable or disable the application of exclusion.
+  virtual function void enable_excl(string obj, bit enable = 1'b1);
+    string index;
+    if (get_excl_index(obj, index)) begin
```

should we use `DV_CHECK to ensure it returns the index?

---

See description above - I intentionally chose to vacuously return rather than treat it as an error case. If the goal is to disable exclusions on ALL registers temporarily, the user can simply loop through all available registers and flip this bit. Otherwise, they will need to check first, which registers have exclusions and only flip the bit for those, which I felt would make the code a bit more tedious. If the exclusions don't exist in the first place, then the attempt to flip the enable bit should have no effect (and this would thus be WAI). 

Alternately, there is a possibility that the user tried to disable the exclusion on a register/field but due to a mistake somewhere, nothing happened because we chose to silently return. 

I decided to add a bit `throw_error` to let the user decide!

---

I see. sounds good.